meerkat/ops/map.py

```diff
@@ -425,23 +430,77 @@ def map(
         output_type=output_type,
         materialize=materialize,
     )
-    return _materialize(deferred, batch_size=batch_size, pbar=pbar, use_ray=use_ray)
+    return _materialize(
+        deferred,
+        batch_size=batch_size,
+        pbar=pbar,
+        use_ray=use_ray,
+        num_blocks=num_blocks,
+        blocks_per_window=blocks_per_window,
+    )
 
 
 def _materialize(
-    data: Union["DataFrame", "Column"], batch_size: int, pbar: bool, use_ray: bool
+    data: Union["DataFrame", "Column"],
+    batch_size: int,
+    pbar: bool,
+    use_ray: bool,
+    num_blocks: int,
+    blocks_per_window: int,
 ):
+    import ray
     from tqdm import tqdm
+    import pandas as pd
+    import numpy as np
+    import meerkat as mk
 
     from .concat import concat
 
     if use_ray:
-        # TODO (dean): Implement this with ray for linear pipelines only, if there are
-        # branches raises a valueerror.
-        # Build the pipeline by following `data.args` and `data.kwargs`
-        # `out = df.defer(lambda img: np.array(img.resize((100, 100)))).defer(lambda img: (img.mean(), img.std()))`
-        # `out["0"].data.args[0].data.kwargs["img"]`
-        raise NotImplementedError
+        ray.init(ignore_reinit_error=True)
+
+        # Step 1: Walk through the DeferredColumns and build a list of functions
+        curr = data
+        fns = []
+        while isinstance(curr, mk.DeferredColumn):
+            fns.append(curr.data.fn)
+
+            # For linear pipelines, there will be either one elem in args or one key in kwargs
+            if curr.data.args:
+                if len(curr.data.args) > 1:
+                    raise ValueError("Multiple args not supported.")
+                curr = curr.data.args[0]
+            elif curr.data.kwargs:
+                if len(curr.data.kwargs) > 1:
+                    raise ValueError("Multiple kwargs not supported.")
+                curr = curr.data.kwargs[next(iter(curr.data.kwargs))]
+            else:
+                raise ValueError("No args or kwargs.")
+
+        # Step 2: Create the ray dataset from the base column
+        if isinstance(curr, mk.ScalarColumn):
+            ds = ray.data.from_pandas(pd.DataFrame({"0": curr})).repartition(num_blocks)
+            fns.append(lambda x: x["0"])
+        else:
+            raise ValueError(f"Base column is of unsupported type {type(curr)}.")
+
+        # Step 3: Build the pipeline by walking backwards through fns
+        pipe: ray.data.DatasetPipeline = ds.window(blocks_per_window=blocks_per_window)
+        for fn in reversed(fns):
+            pipe = pipe.map(fn)
+
+        # Step 4: Collect the results
+        # TODO (dean): support different output types
+        result = np.array([])
```

Comment:
Using append works in the scalar case, but if the returned function is an embedding, then a  np.stack or np.concatenate may be needed. In general, can be good to accumulate the array in a python list and then stack or concat after, since appends to np arrays involve a full copy of the array. 

You can test the new functionality on this script: 
```
import meerkat as mk
import torch
import clip
from PIL import Image


df = mk.get("imagenette")[:100]

device = "cuda" if torch.cuda.is_available() else "cpu"
model, preprocess = clip.load("ViT-B/32", device=device)

@torch.no_grad()
def embed_image(img: torch.tensor):
    return model.encode_image(img.unsqueeze(0).to(device)).cpu().detach().numpy().squeeze()

df["preprocessed"] = df.defer(
    lambda img: preprocess(img)
)

df["emb"] = df.map(
    lambda preprocessed: embed_image(preprocessed),
    use_ray=True
)
```